services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/utils/GoalStateUtil.java

```diff
@@ -0,0 +1,158 @@
+                final PortState portState1 =
+                        PortState.newBuilder().setConfiguration(portConfiguration).build();
+                PortState portState = AssignVipMacToPort(portState1,
+                        random.nextInt(Integer.MAX_VALUE));
+                portStateArrayList.add(portState);
+//TODO would add fields transformation 1 by 1
+//                    .setFixedIps(e.getFixedIps()) // we need an array
+            });
+            final Goalstate.GoalState goalState =
+                    Goalstate.GoalState.newBuilder().addAllPortStates(portStateArrayList).build();
+            goalStateHashMap.put(f.getKey(), goalState);
+
+
+        });
+
+
+        return goalStateHashMap;
+    }
+
+    public static List<Goalstateprovisioner.GoalStateOperationReply> talkToACA(Map<String, Goalstate.GoalState> gss, boolean isFast) {
+        //if Config.isOVS
+        GoalStateService goalStateService = new OVSGoalStateServiceImpl();
+
+        return gss.entrySet().parallelStream().map(e -> {
+            goalStateService.setIp(e.getKey());
+            goalStateService.setGoalState(e.getValue());
+            goalStateService.setFastPath(isFast);
+            return goalStateService.SendGoalStateToHosts();
+        }).collect(Collectors.toList());
+    }
+
+    public static PortState AssignVipMacToPort(PortState portState,
+                                               int epIndex) {
+
+        Port.PortConfiguration.Builder conf =
+                portState.getConfiguration().newBuilder();
+        conf.setMacAddress(GenerateMacAddress(epIndex));
+        String[] vpcIps = new String[]{GenerateIpAddress(epIndex)};
+        try {
+
+            List<FixedIp> fixedIps = convertToFixedIps(vpcIps,
+                    portState.getConfiguration().getNetworkId());
+            for (FixedIp fip : fixedIps) {
+                conf.addFixedIps(fip);
+            }
+        } catch (InvalidProtocolBufferException e) {
+            e.printStackTrace();
+        }
+        PortState.Builder builder = portState.newBuilderForType();
+        builder.setConfiguration(conf);
+        return builder.build();
+    }
+
+    public static List<FixedIp> convertToFixedIps(String[] vpcIps,
+                                                  String subnetId) throws InvalidProtocolBufferException {
+
+        List<FixedIp> fixedIps = new ArrayList<>();
+        if (vpcIps != null) {
+            for (String vpcIp : vpcIps) {
+                FixedIp.Builder builder = FixedIp.newBuilder();
+                FixedIp build =
+                        builder.setSubnetId(subnetId).setIpAddress(vpcIp).build();
+                fixedIps.add(build);
+            }
+        }
+
+        return fixedIps;
+    }
+
+    private static String GenerateMacAddress(int index) {
+        return "0e:73:ae:c8:" + Integer.toHexString((index + 6) / 256) + ":" + Integer.toHexString((index + 6) % 256);
+    }
+
+    private static String GenerateIpAddress(int index) {
```

Comment:
we need to get a better virtual IP address either by:
1) input by customer
2) get a valid (and unique) one with the subnet